File: web/src/soma/presets.ts
// Tracker presets — port of soma/cli.py VARIANTS + PRESETS.
// SOMA: geometry/parts/orientation channels only (no ReID).
// SOMA-R: appearance-dominant stage-1 + memory/revival/ghost stack; the
// embedding thresholds are quantile-calibrated per ReID model.

import type { TrackerConfig } from './tracker';

export type VariantId = 'det' | 'pv' | 'os';

const SOMAR_BASE: Partial<TrackerConfig> = {
  detThresh: 0.35,
  initThresh: 0.55,
  simGate: 0.3,
  wEmb: 4.0,
  wOks: 0.0,
  amodalAlpha: 4.0,
  dloBeta: 0.65,
  tokenFloor: 0.12,
  wDir: 0.0,
  dirVeto: 0.0,
  sizePriorThr: 0.5,
  embCenterLambda: 0.25,
  memTtlSec: 12.0,
  memMargin: 0.1,
  lostVelDecay: 0.85,
  gateGrow: 0.6,
  maxAgeSec: 6.0,
  lostEmbGateSec: 1.2,
  emitKf: true,
  partGate: 1.01,
  embUpdateCrowdMax: 0.6,
  ghostEmitMaxS: 0.1,
  ghostCrowdMax: 0.3,
};

export const PRESETS: Record<string, Partial<TrackerConfig>> = {
  soma: {
    detThresh: 0.45,
    initThresh: 0.55,
    simGate: 0.2,
    amodalAlpha: 4.0,
    dloBeta: 0.65,
    tokenFloor: 0.25,
    wDir: 0.5,
    dirVeto: 0.15,
    sizePriorThr: 0.5,
  },
  'somar-pv': {
    ...SOMAR_BASE,
    embLo: 0.398,
    embHi: 0.945,
    embVeto: 0.479,
    lostEmbGateCos: 0.398,
    memCos: 0.642,
    reviveCos: 0.56,
  },
  'somar-os': {
    ...SOMAR_BASE,
    embLo: 0.25,
    embHi: 0.899,
    embVeto: 0.314,
    lostEmbGateCos: 0.25,
    memCos: 0.475,
    reviveCos: 0.384,
  },
};

export interface VariantInfo {
  id: VariantId;
  label: string;
  preset: keyof typeof PRESETS;
  usesReid: boolean;
  // causal whitening EMA keep-rate (soma/cli.py VARIANTS whiten)
  whiten: number;
}

export const VARIANTS: VariantInfo[] = [
  { id: 'det', label: 'SOMA (no ReID)', preset: 'soma', usesReid: false, whiten: 0.0 },
  { id: 'pv', label: 'SOMA-R / PersonViT (raw)', preset: 'somar-pv', usesReid: true, whiten: 0.0 },
  { id: 'os', label: 'SOMA-R / OSNet-AIN (whitened)', preset: 'somar-os', usesReid: true, whiten: 0.98 },
];

// perception defaults used by the python cache/video paths
export const BODY_SCORE_FLOOR = 0.1;

File: soma/presets.py
"""Shipped tracker presets and ReID variants (the public configuration API).

SOMA: geometry/parts/orientation channels only (no ReID).
SOMA-R: appearance-dominant stage-1 + memory/revival/ghost stack; the
embedding thresholds are quantile-calibrated per ReID model (ptrack exp050
procedure), everything else is shared.
"""
from __future__ import annotations

from .tracker import TrackerConfig

# ReID variants of the wb28x640 CrowdTrack benchmark. PersonViT v3 runs RAW
# (its raw cosine geometry is already spread — internal domain
# normalization); OSNet-AIN aug v3 needs causal whitening (compressed cone).
VARIANTS = {
    "det": {"reid": None, "whiten": 0.0},
    "pv": {"reid": "models/personvit_vits16_ain_unified_aug_n.onnx", "whiten": 0.0},
    "os": {"reid": "models/osnet_ain_x1_0_p_unified_aug_n.onnx", "whiten": 0.98},
}

_SOMAR_BASE = {
    "det_thresh": 0.35, "init_thresh": 0.55, "sim_gate": 0.30,
    "w_emb": 4.0, "w_oks": 0.0, "amodal_alpha": 4.0, "dlo_beta": 0.65,
    "token_floor": 0.12, "w_dir": 0.0, "dir_veto": 0.0, "size_prior_thr": 0.5,
    "emb_center_lambda": 0.25, "mem_ttl_sec": 12.0, "mem_margin": 0.10,
    "lost_vel_decay": 0.85, "gate_grow": 0.6, "max_age_sec": 6.0,
    "lost_emb_gate_sec": 1.2, "emit_kf": True, "part_gate": 1.01,
    "emb_update_crowd_max": 0.60,
    "ghost_emit_max_s": 0.10, "ghost_crowd_max": 0.30,
}
PRESETS = {
    "soma": {"det_thresh": 0.45, "init_thresh": 0.55, "sim_gate": 0.20,
             "amodal_alpha": 4.0, "dlo_beta": 0.65, "token_floor": 0.25,
             "w_dir": 0.5, "dir_veto": 0.15, "size_prior_thr": 0.5},
    "somar-pv": {**_SOMAR_BASE,
                 "emb_lo": 0.398, "emb_hi": 0.945, "emb_veto": 0.479,
                 "lost_emb_gate_cos": 0.398, "mem_cos": 0.642,
                 "revive_cos": 0.560},
    # v4 (cam-branch generation, ptrack exp065): whitening kept — the raw
    # pool is healthy now but whitening still wins the cell A/B; thresholds
    # re-qmapped on the v4 whitened pool.
    "somar-os": {**_SOMAR_BASE,
                 "emb_lo": 0.251, "emb_hi": 0.901, "emb_veto": 0.314,
                 "lost_emb_gate_cos": 0.251, "mem_cos": 0.470,
                 "revive_cos": 0.383},
}


def tracker_config(preset: str = "somar-pv", fps: int = 30,
                   **overrides) -> TrackerConfig:
    """Build a ready-to-run TrackerConfig from a shipped preset.

    Resolves the frame-rate-derived horizons (``max_age`` from
    ``max_age_sec``) so callers never repeat the driver boilerplate.
    ``overrides`` are applied on top of the preset fields.
    """
    if preset not in PRESETS:
        raise ValueError(f"unknown preset: {preset!r}; expected one of: "
                         f"{', '.join(sorted(PRESETS))}")
    kw = dict(PRESETS[preset])
    kw.update(overrides)
    cfg = TrackerConfig(**kw)
    cfg.fps = int(fps)
    cfg.max_age = (int(round(cfg.max_age_sec * cfg.fps))
                   if cfg.max_age_sec > 0 else 2 * cfg.fps)
    return cfg
